### PR TITLE
STORM-1340: Use Travis-CI build matrix to improve test execution times

### PR DESCRIPTION
**Pull request by knusbaum**
This patch modifies the Travis-CI config to run storm-core unit tests in parallel with all the external tests, knocking usually between 4 and 7 minutes off the build time usually.

It also caches ruby, javascript, and java libraries instead of downloading them each time a container is launched.

### CODE REVIEW

File: dev-tools/travis/travis-install.sh [hunk 1]
```diff
@@ -29,6 +29,7 @@ TRAVIS_SCRIPT_DIR=$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )
 cd ${STORM_SRC_ROOT_DIR}
 
 python ${TRAVIS_SCRIPT_DIR}/save-logs.py "install.txt" mvn clean install -DskipTests -Pnative --batch-mode
+
```
Comment (d2r) [inline dev-tools/travis/travis-install.sh]: revert?

**Comment by revans2** [comment]
Looks good to me drops the build time by about 7 mins. I am +1 if you can address @d2r comment and squash yur changes.  11 commits feels like a bit much.

**Comment by knusbaum** [comment]
Failure looks unrelated.

**Comment by d2r** [comment]
Looks good to me +1

**Comment by knusbaum** [comment]
@HeartSaVioR Would you like to look at this? 
It's similar to the PR I put up before.

**Comment by HeartSaVioR** [comment]
@knusbaum 
I'm +1 for the change.
Actually I found some builds before rebase still takes 13 min or higher, but mostly it seems to save build time.
Great work!

**Comment by Taher-Ghaleb** [comment]
Hello. Sorry, it's been a long time ago, but I am curious why caching hasn't been enabled before this (i.e., since December 17, 2014 when it became available to open source projects). Thank you.

**Comment by knusbaum** [comment]
@Taher-Ghaleb sorry, I'm not sure I understand your question. Would you mind clarifying?

**Comment by Taher-Ghaleb** [comment]
Hi @knusbaum,
Thanks for your reply.

I see you have enabled `cache:` for Travis builds in this pull request, whereas the [caching feature of Travis was available since December 17, 2014](https://blog.travis-ci.com/2014-12-17-faster-builds-with-container-based-infrastructure) (about a year earlier). I am interested in understanding the reason behind delaying the `cache` feature.

Thank you.

**Comment by knusbaum** [comment]
@Taher-Ghaleb
At this point I have little memory of this issue. My guess is that it was not a priority for our team until build times started getting very long.

**Comment by Taher-Ghaleb** [comment]
I see your point. Thank you so much, @knusbaum, for your gentle responsiveness.

**Comment by d2r** [comment]
> > I am interested in understanding the reason behind delaying the `cache` feature.

> it was not a priority for our team until build times started getting very long.

@Taher-Ghaleb That is my memory as well. The [jira issue](https://issues.apache.org/jira/browse/STORM-1340) was not created until Nov 2015, and that was likely the first time we decided to address it.

**Comment by Taher-Ghaleb** [comment]
OK, I see. Thanks, @d2r, for such details.

**Pull request merged**